cng/mlkem.go

```diff
@@ -0,0 +1,449 @@
+}
+
+// generateMLKEMKey generates a new ML-KEM key pair with the specified parameter set
+// and writes the raw key bytes (not the blob) into dst.
+func generateMLKEMKey(paramSet string, dst []byte) error {
+	alg, err := loadMLKEM()
+	if err != nil {
+		return err
+	}
+
+	var hKey bcrypt.KEY_HANDLE
+	err = bcrypt.GenerateKeyPair(alg.handle, &hKey, 0, 0)
+	if err != nil {
+		return errors.New("mlkem: key generation failed")
+	}
+	defer bcrypt.DestroyKey(hKey)
+
+	// Set the parameter set
+	if err := setString(bcrypt.HANDLE(hKey), bcrypt.PARAMETER_SET_NAME, paramSet); err != nil {
+		return err
+	}
+	if err != nil {
+		return errors.New("mlkem: failed to set parameter set")
+	}
+
+	err = bcrypt.FinalizeKeyPair(hKey, 0)
+	if err != nil {
+		return errors.New("mlkem: key finalization failed")
+	}
+
+	// Export the private key blob
+	var size uint32
+	err = bcrypt.ExportKey(hKey, 0, utf16PtrFromString(bcrypt.MLKEM_PRIVATE_BLOB), nil, &size, 0)
+	if err != nil {
+		return errors.New("mlkem: failed to get key blob size")
+	}
+
+	blob := make([]byte, size)
+	err = bcrypt.ExportKey(hKey, 0, utf16PtrFromString(bcrypt.MLKEM_PRIVATE_BLOB), blob, &size, 0)
+	if err != nil {
+		return errors.New("mlkem: failed to export key")
+	}
+
+	// Extract raw key bytes into destination
+	return extractMLKEMKeyBytes(blob, dst)
+}
+
+// newMLKEMKeyBlob creates a key blob from raw key bytes.
+func newMLKEMKeyBlob(paramSet string, keyBytes []byte, magic bcrypt.KeyBlobMagicNumber) ([]byte, error) {
+	paramSetUTF16 := utf16FromString(paramSet)
+	paramSetByteLen := len(paramSetUTF16) * 2
+
+	blob := make([]byte, 12+paramSetByteLen+len(keyBytes))
+	putUint32LE(blob[0:4], uint32(magic))
+	putUint32LE(blob[4:8], uint32(paramSetByteLen)) // cbParameterSet
+	putUint32LE(blob[8:12], uint32(len(keyBytes)))  // cbKey
+	for i, v := range paramSetUTF16 {
+		putUint16LE(blob[12+i*2:], v)
+	}
+	copy(blob[12+paramSetByteLen:], keyBytes)
+
+	return blob, nil
+}
+
+// extractMLKEMKeyBytes extracts the raw key bytes from a blob into the provided destination slice.
+func extractMLKEMKeyBytes(blob []byte, dst []byte) error {
+	if len(blob) < 12 {
+		return errors.New("mlkem: blob too small")
+	}
+	cbParameterSet := getUint32LE(blob[4:8])
+	cbKey := getUint32LE(blob[8:12])
+	headerSize := 12 + int(cbParameterSet)
+	if len(blob) < headerSize+int(cbKey) {
+		return errors.New("mlkem: invalid blob size")
+	}
+	if len(dst) != int(cbKey) {
+		return errors.New("mlkem: destination size mismatch")
+	}
+	copy(dst, blob[headerSize:headerSize+int(cbKey)])
+	return nil
+}
+
+// mlkemDecapsulate is a shared helper for decapsulating with ML-KEM keys.
+func mlkemDecapsulate(paramSet string, keyBytes []byte, ciphertext []byte, expectedCiphertextSize int) ([]byte, error) {
+	if len(ciphertext) != expectedCiphertextSize {
+		return nil, errors.New("mlkem: invalid ciphertext size")
+	}
+
+	alg, err := loadMLKEM()
+	if err != nil {
+		return nil, errors.New("mlkem: failed to open algorithm provider")
```

Comment:
return `err` rather than a custom error string. Do the same for other error cases.